libs/libbuiltin/libgcc/gcov.c

```diff
@@ -0,0 +1,272 @@
+  unsigned int cfg_checksum;       /* Function cfg checksum */
+  struct gcov_ctr_info ctrs[1];    /* Instrumented counters */
+};
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+FAR struct gcov_info *__gcov_info_start;
+FAR struct gcov_info *__gcov_info_end;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void dump_counter(FAR void *buffer, FAR size_t *off, uint64_t v)
+{
+  if (buffer)
+    {
+      memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
+    }
+
+  *off += sizeof(uint64_t);
+}
+
+static void dump_unsigned(FAR void *buffer, FAR size_t *off, uint32_t v)
+{
+  if (buffer)
+    {
+      memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
+    }
+
+  *off += sizeof(uint32_t);
+}
+
+static size_t gcov_convert(FAR uint8_t *buffer,
+                           FAR const struct gcov_info *info)
+{
+  FAR struct gcov_fn_info *gfi;
+  FAR struct gcov_ctr_info *gci;
+  uint32_t functions;
+  uint32_t counts;
+  uint32_t value;
+  size_t pos = 0;
+
+  /* File header. */
+
+  dump_unsigned(buffer, &pos, GCOV_DATA_MAGIC);
+  dump_unsigned(buffer, &pos, info->version);
+  dump_unsigned(buffer, &pos, info->stamp);
+
+#ifdef GCOV_12_FORMAT
+  dump_unsigned(buffer, &pos, info->checksum);
+#endif
+
+  /* Function headers. */
+
+  for (functions = 0; functions < info->n_functions; functions++)
+    {
+      gfi = info->functions[functions];
+
+      /* Function record. */
+
+      dump_unsigned(buffer, &pos, GCOV_TAG_FUNCTION);
+      dump_unsigned(buffer, &pos, GCOV_TAG_FUNCTION_LENGTH);
+      dump_unsigned(buffer, &pos, gfi->ident);
+      dump_unsigned(buffer, &pos, gfi->lineno_checksum);
+      dump_unsigned(buffer, &pos, gfi->cfg_checksum);
+
+      gci = gfi->ctrs;
+      for (counts = 0; counts < GCOV_COUNTERS; counts++)
+        {
+          if (!info->merge[counts])
+            {
+              continue;
+            }
+
+          /* Counter record. */
+
+          dump_unsigned(buffer, &pos, GCOV_TAG_FOR_COUNTER(counts));
+          dump_unsigned(buffer, &pos, gci->num * 2 * GCOV_UNIT_SIZE);
+
+          for (value = 0; value < gci->num; value++)
+            {
+              dump_counter(buffer, &pos, gci->values[value]);
+            }
+
+          gci++;
+        }
+    }
+
+  return pos;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void __gcov_init(FAR struct gcov_info *info)
+{
+  info->next = __gcov_info_start;
+  __gcov_info_start = info;
+}
+
+void __gcov_merge_add(FAR gcov_type *counters, unsigned int n_counters)
+{
+}
+
+void __gcov_exit(void)
+{
+}
+
+void __gcov_dump(void)
```

Comment:
why not implement __gcov_dump

Reply:
but this is mini version, how can we enable toolchain and mini version at the same time?

Reply:
Done